opengovsg/GoGovSG
# Release 1.70

## Tests

## Home page
- [x] The landing page graphic should animate example short urls correctly
- [x] The landing page should have no link to search

## Login page
- [x] OTP email should contain requestor's IP address
- [x] Resend OTP should send a new OTP to user, and invalidate previous OTP

## User page
- [x] User message banner appears on top when USER_MESSAGE env var is set
- [x] Announcement modal appears when ANNOUNCEMENT_* env var is set
- [x] Ellipses are shown for very long links
- [x] Banner follows scroll

## URL creation
- [x] It should prevent uploading when a [malicious file](https://www.webroot.com/blog/2018/09/05/eicar-common-false-positive-world/) is submitted

## URL Searching and Download
- [x] Clicking on the download button downloads all links currently shown in the links table as a `.csv` file.

## Redirection
- [x] It should redirect to the long url when short url is active
- [x] It should redirect to a 404 page when short url is inactive

## Drawer, Editing and Toaster logic
- [x] It should redirect to the amended file immediately when the file is replaced (any caching for that short url is cleared)

## QR
- [x] Downloaded PNG QR code should be of width 1000px and height >= 1000px
- [x] Downloaded SVG QR code should work
- [x] Scan the downloaded QR code in SVG to check that it points to the correct long url
- [x] Scan the downloaded QR code in PNG to check that it points to the correct long url

## User page - Sort & Filter
- [x] Links should be sorted by their number of clicks in descending order when enabling sort by `Most number of visits` and clicking apply

## Sequelize hooks
- [x] UrlHistory table is populated in database

## Directory - desktop view
- [x] It should redirect to the short url link when the row is clicked for active link

## Directory - mobile view
- [x] It should redirect to the short url link when the redirect icon is clicked for active link in bottom mobile panel

## Environment variables

- `BULK_UPLOAD_MAX_NUM` | Required? - No | Maximum number of links that can be bulk uploaded at once. Defaults to 1000|
- `BULK_UPLOAD_RANDOM_STR_LENGTH` | Required? - No | String length of randomly generated shortUrl in bulk upload. Defaults to 8|


## Migration Notes

- Run migration `20220929_add_source_column.sql` on Go, Edu, Health
- Verify nothing is breaking
  - Test link creation, retrieval, link update
- Deploy by merging branch `release-1.70` to `release`
- Verify nothing is breaking
  - Test link creation, retrieval, link update
  - Source column should be written into
  - Bulk creation also adds rows in url clicks, url history
- Add env var` BULK_UPLOAD_MAX_NUM=50000` to Go environment
- Migrate Serverless link transfer functions on Go, Edu, Health
- Increase invocation time for migrate_user_links from 6s to 1 min

Off peak:
- Run migration 20220929_backfill_source_column.sql on Go, Edu, Health
- Verify nothing is breaking
  - Test link creation, retrieval, link update
  - Source column should be backfilled
  - Link transfers and link updates should work
